### PR TITLE
Stepper: Fix back button on use my domain step

### LINKED ISSUE #94234: Stepper: Domains step follow up
### Details

- `Back` in `use-my-domain` doesn't work. It returns you to domains, while it should bring you to the input in the screen before

image

### PR DESCRIPTION
**Pull request by agrullon95**
Fixes: https://github.com/Automattic/wp-calypso/issues/94234

## Proposed Changes

* Updates the back button navigation for the use my domain step

## Why are these changes being made?

* Allow users to select a choose a different domain instead of navigating back to domains and then navigating to use my domain step again.

## Testing Instructions

* Checkout branch or use live link
* Navigate to `/setup/onboarding` and click the `Use a domain I own` link on the side bar.
* Continue through the flow until you reach the checkout and verify that the correct domain cart items are added
* Test the back navigation
* Check for regressions on `/start/` flow.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

**Comment by renancarvalho** [comment]
Thank yo for working on this.

I noticed one thing when comparing against start.

When removing the selected plan from the checkout, using setup I get redirected to `/plans` in calypso, on the other hand, using start, I got redirected to `https://wpcalypso.wordpress.com/setup/site-setup/goals?siteId=236827050&skippedCheckout=1` using live link. I am unsure if this is related to your changes. Just noting here.

File: client/components/domains/use-my-domain/index.jsx [hunk 1]
```diff
@@ -95,6 +92,10 @@ function UseMyDomain( props ) {
 					setOwnershipVerificationFlowPageSlug( prevOwnershipVerificationFlowPageSlug );
 				} else {
 					setMode( inputMode.transferOrConnect );
+
+					if ( isStepper ) {
+						setUseMyDomainMode?.( inputMode.transferOrConnect );
```
Comment (renancarvalho) [inline client/components/domains/use-my-domain/index.jsx]: What if we extract this repetition into a function and reuse it?

Something like:
```js

 const setModeAndStepper = (newMode) => {
        setMode(newMode);
        if (isStepper) {
            setUseMyDomainMode?.(newMode);
        }
    };

```

File: client/components/domains/use-my-domain/index.jsx [hunk 2]
```diff
@@ -422,6 +431,36 @@ function UseMyDomain( props ) {
...
+		}
+	}, [ initialMode ] );
+
+	useEffect( () => {
+		if ( ! initialQuery || initialValidation.current ) {
+			return;
+		}
+
+		initialValidation.current = true;
+		initialQuery &&
```
Comment (renancarvalho) [inline client/components/domains/use-my-domain/index.jsx:435]: Did not get this part, why do we need it?
`initialQuery && ! initialMode && ! getDomainNameValidationErrorMessage( initialQuery ) && onNext();`

File: client/landing/stepper/declarative-flow/internals/steps-repository/use-my-domain/index.tsx [hunk 3]
```diff
@@ -47,7 +61,9 @@ const UseMyDomain: Step = function UseMyDomain( { navigation, flow } ) {
 	};
 
 	const getInitialQuery = function () {
-		return String( getQueryArg( window.location.search, 'lastQuery' ) ?? '' );
+		const lastQuery = String( getQueryArg( window.location.search, 'lastQuery' ) ?? '' );
+		const initialQuery = String( getQueryArg( window.location.search, 'initialQuery' ) ?? '' );
+		return lastQuery || initialQuery || '';
```
Comment (renancarvalho) [inline client/landing/stepper/declarative-flow/internals/steps-repository/use-my-domain/index.tsx]: I think `return lastQuery || initialQuery` is enough.

File: client/landing/stepper/declarative-flow/internals/steps-repository/use-my-domain/index.tsx [hunk 4]
```diff
@@ -78,13 +97,21 @@ const UseMyDomain: Step = function UseMyDomain( { navigation, flow } ) {
 		}
 	};
 
+	useEffect( () => {
+		const stepQueryParam = getQueryArg( window.location.search, 'step' );
+		if ( stepQueryParam === 'transfer-or-connect' && stepQueryParam !== useMyDomainMode ) {
+			setUseMyDomainMode( inputMode.transferOrConnect );
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [] );
```
Comment (renancarvalho) [inline client/landing/stepper/declarative-flow/internals/steps-repository/use-my-domain/index.tsx]: Don't we need `useMyDomainMode` or `inputMode` as a dependency?

File: client/landing/stepper/declarative-flow/onboarding.ts [hunk 5]
```diff
@@ -103,11 +108,23 @@ const onboarding: Flow = {
 		};
 
 		const goBack = () => {
+			const lastQuery = getQueryArg( window.location.href, 'lastQuery' );
+
 			switch ( currentStepSlug ) {
 				case 'use-my-domain':
+					if ( lastQuery ) {
+						// remove query params
+						// window.history.pushState( {}, document.title, window.location.pathname );
```
Comment (renancarvalho) [inline client/landing/stepper/declarative-flow/onboarding.ts]: I Think we can remove the commented code.

**Review by renancarvalho — COMMENTED** [review]
Left some comments :)

File: client/components/domains/use-my-domain/index.jsx (hunk 2 above)
Comment (agrullon95) [reply client/components/domains/use-my-domain/index.jsx:435]: This is existing logic. I just moved it to the bottom of the file.

**Comment by renancarvalho** [comment]
Found out this one, 
https://github.com/user-attachments/assets/ac89feae-508d-43bf-b6c1-b6d790bd9c40


Also, I realized the browser back button is always pointing back to use-my-domain my domain step, is it intended?

**Review by escapemanuele — CHANGES_REQUESTED** [review]
- /start has a header in `use-your-domain`

image

- Browser back button does not work in `transfer-or-connect` step

- When reaching the plans step after clicking `Select` on `Transfer your domain` here

image

I should go back to `step=transfer-or-connect`, instead I return to the `use-your-domain` input page

- Plans step Back button redirects to domains instead of `use-your-domain`

**Comment by renancarvalho** [comment]
Good work @agrullon95 Navigation using the back button seems fine 👍 

Just noted:

On live link -> /start it looks broken

image

Also, the translations are not correct 

Stepper
image

Start in production
image

**Comment by renancarvalho** [comment]
Havent found more issues, I think we are good to go.

**Review by renancarvalho — APPROVED** [review]

**Pull request merged**